datafusion/expr/src/binary_rule.rs

```diff
@@ -185,6 +186,17 @@ fn comparison_order_coercion(
         .or_else(|| null_coercion(lhs_type, rhs_type))
 }
 
+fn string_numeric_coercion(lhs_type: &DataType, rhs_type: &DataType) -> Option<DataType> {
+    use arrow::datatypes::DataType::*;
+    match (lhs_type, rhs_type) {
```

Comment:
I test in `748b6a65a5fa801595fd80a3c7b2728be3c9cdaa`(not this commit)

```
explain select * from part where p_partkey in (1, 2, '3');
+---------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| plan_type     | plan                                                                                                                                                                                                                                                                      |
+---------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| logical_plan  | Projection: #part.p_partkey, #part.p_name, #part.p_mfgr, #part.p_brand, #part.p_type, #part.p_size, #part.p_container, #part.p_retailprice, #part.p_comment                                                                                                               |
|               |   Filter: #part.p_partkey IN ([Int64(1), Int64(2), Utf8("3")])                                                                                                                                                                                                            |
|               |     TableScan: part projection=Some([p_partkey, p_name, p_mfgr, p_brand, p_type, p_size, p_container, p_retailprice, p_comment]), partial_filters=[#part.p_partkey IN ([Int64(1), Int64(2), Utf8("3")])]                                                                  |
| physical_plan | ProjectionExec: expr=[p_partkey@0 as p_partkey, p_name@1 as p_name, p_mfgr@2 as p_mfgr, p_brand@3 as p_brand, p_type@4 as p_type, p_size@5 as p_size, p_container@6 as p_container, p_retailprice@7 as p_retailprice, p_comment@8 as p_comment]                           |
|               |   CoalesceBatchesExec: target_batch_size=4096                                                                                                                                                                                                                             |
|               |     FilterExec: p_partkey@0 IN ([Literal { value: Int64(1) }, Literal { value: Int64(2) }, CastExpr { expr: Literal { value: Utf8("3") }, cast_type: Int64, cast_options: CastOptions { safe: false } }])                                                                 |
|               |       RepartitionExec: partitioning=RoundRobinBatch(16)                                                                                                                                                                                                                   |
|               |         ParquetExec: limit=None, partitions=[/Users/yangjiang/test-data/tpch-1g-oneFile/part/part-00000-3a3c2777-00d3-4c27-b917-4ff2145123dc-c000.snappy.parquet], projection=[p_partkey, p_name, p_mfgr, p_brand, p_type, p_size, p_container, p_retailprice, p_comment] |
|               |                                                                                                                                                                                                                                                                           |
+---------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

`int, int,utf8` cast to -> `int, int, int`,

In my opinion, after apply this patch it will get   int, int,utf8` cast to -> `utf8, utf8, utf8`
I think when list_values_size is large, we will construct a hashSet in https://github.com/apache/arrow-datafusion/pull/2156,  change to `int` will get better performance in build hasSet, Am i right?  😄

Reply:
test in this patch
```
explain select * from part where p_partkey in (1, 2, '3');
+---------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| plan_type     | plan                                                                                                                                                                                                                                                                      |
+---------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| logical_plan  | Projection: #part.p_partkey, #part.p_name, #part.p_mfgr, #part.p_brand, #part.p_type, #part.p_size, #part.p_container, #part.p_retailprice, #part.p_comment                                                                                                               |
|               |   Filter: #part.p_partkey IN ([Int64(1), Int64(2), Utf8("3")])                                                                                                                                                                                                            |
|               |     TableScan: part projection=Some([p_partkey, p_name, p_mfgr, p_brand, p_type, p_size, p_container, p_retailprice, p_comment]), partial_filters=[#part.p_partkey IN ([Int64(1), Int64(2), Utf8("3")])]                                                                  |
| physical_plan | ProjectionExec: expr=[p_partkey@0 as p_partkey, p_name@1 as p_name, p_mfgr@2 as p_mfgr, p_brand@3 as p_brand, p_type@4 as p_type, p_size@5 as p_size, p_container@6 as p_container, p_retailprice@7 as p_retailprice, p_comment@8 as p_comment]                           |
|               |   CoalesceBatchesExec: target_batch_size=4096                                                                                                                                                                                                                             |
|               |     FilterExec: CAST(p_partkey@0 AS Utf8) IN ([TryCastExpr { expr: Literal { value: Int64(1) }, cast_type: Utf8 }, TryCastExpr { expr: Literal { value: Int64(2) }, cast_type: Utf8 }, Literal { value: Utf8("3") }])                                                     |
|               |       RepartitionExec: partitioning=RoundRobinBatch(16)                                                                                                                                                                                                                   |
|               |         ParquetExec: limit=None, partitions=[/Users/yangjiang/test-data/tpch-1g-oneFile/part/part-00000-3a3c2777-00d3-4c27-b917-4ff2145123dc-c000.snappy.parquet], projection=[p_partkey, p_name, p_mfgr, p_brand, p_type, p_size, p_container, p_retailprice, p_comment] |
|               |                                                                                                                                                                                                                                                                           |
+---------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

Reply:
As it is somehow following current behavior, I can address it in the other issue #2799.

Reply:
> I test in `748b6a65a5fa801595fd80a3c7b2728be3c9cdaa`(not this commit)
> 
> ```
> explain select * from part where p_partkey in (1, 2, '3');
> +---------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
> | plan_type     | plan                                                                                                                                                                                                                                                                      |
> +---------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
> | logical_plan  | Projection: #part.p_partkey, #part.p_name, #part.p_mfgr, #part.p_brand, #part.p_type, #part.p_size, #part.p_container, #part.p_retailprice, #part.p_comment                                                                                                               |
> |               |   Filter: #part.p_partkey IN ([Int64(1), Int64(2), Utf8("3")])                                                                                                                                                                                                            |
> |               |     TableScan: part projection=Some([p_partkey, p_name, p_mfgr, p_brand, p_type, p_size, p_container, p_retailprice, p_comment]), partial_filters=[#part.p_partkey IN ([Int64(1), Int64(2), Utf8("3")])]                                                                  |
> | physical_plan | ProjectionExec: expr=[p_partkey@0 as p_partkey, p_name@1 as p_name, p_mfgr@2 as p_mfgr, p_brand@3 as p_brand, p_type@4 as p_type, p_size@5 as p_size, p_container@6 as p_container, p_retailprice@7 as p_retailprice, p_comment@8 as p_comment]                           |
> |               |   CoalesceBatchesExec: target_batch_size=4096                                                                                                                                                                                                                             |
> |               |     FilterExec: p_partkey@0 IN ([Literal { value: Int64(1) }, Literal { value: Int64(2) }, CastExpr { expr: Literal { value: Utf8("3") }, cast_type: Int64, cast_options: CastOptions { safe: false } }])                                                                 |
> |               |       RepartitionExec: partitioning=RoundRobinBatch(16)                                                                                                                                                                                                                   |
> |               |         ParquetExec: limit=None, partitions=[/Users/yangjiang/test-data/tpch-1g-oneFile/part/part-00000-3a3c2777-00d3-4c27-b917-4ff2145123dc-c000.snappy.parquet], projection=[p_partkey, p_name, p_mfgr, p_brand, p_type, p_size, p_container, p_retailprice, p_comment] |
> |               |                                                                                                                                                                                                                                                                           |
> +---------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
> ```
> 
> `int, int,utf8` cast to -> `int, int, int`,
> 
> In my opinion, after apply this patch it will get int, int,utf8`cast to ->`utf8, utf8, utf8`I think when list_values_size is large, we will construct a hashSet in https://github.com/apache/arrow-datafusion/pull/2156, change to`int` will get better performance in build hasSet, Am i right? 😄

Yes, the performance is greater for comparing integer.

Now the coercion rule is unstable, we should do mush work in that.

Reply:
> I'm fine if we can get a consensus about if numeric type is more correct for such cases. Then I can change them (the test cases) and the coercion rule.

I think what is important here is to have a consistent set of semantics. I don't have any particular preference related to the automatic coercion of  `int, int, utf8` as I think there are different tradeoffs

For example coercing `int, int, utf8` to `utf8, utf8 utf8` would allow a predicate like `c1 IN (1, 2 'foo')` to run without error (assuming c1 can be coerced to `utf8`), but would result in a runtime error if we attempted to automatically coerce to `int, int int`. However, as @Ted-Jiang notes, the performance will be slower for predicates like `c1 IN (1, 2, '3')`

Best practice would be to explicitly cast all columns to `i32` in the query if they were supposed to be compared as `i32`:

```sql
c1 in (1::smallint, 2::smallint, '3'::smallint)
```

but I realize that may not be practical for all users. 🤔